connaisseur/alert.py

```diff
@@ -0,0 +1,186 @@
+
+    context: dict
+    admission_request: dict
+    throw_if_alert_sending_fails: bool
+
+    def __init__(self, alert_message, receiver_config, admission_request):
+        self.context = {
+            "alert_message": alert_message,
+            "priority": str(receiver_config.get("priority", 3)),
+            "connaisseur_pod_id": os.getenv("POD_NAME"),
+            "cluster": os.getenv("CLUSTER_NAME"),
+            "timestamp": datetime.now(),
+            "request_id": admission_request.get("request", {}).get(
+                "uid", "No given UID"
+            ),
+            "images": (str(get_images(admission_request)) or "No given images"),
+        }
+        self.admission_request = admission_request
+        self.receiver_url = receiver_config["receiver_url"]
+        self.template = receiver_config["template"]
+        self.throw_if_alert_sending_fails = receiver_config.get(
+            "fail_if_alert_sending_fails", False
+        )
+        self.payload = self._construct_payload(receiver_config)
+        self.headers = _get_headers(receiver_config)
+
+    def _construct_payload(self, receiver_config):
+        try:
+            alert_config_dir = f'{os.getenv("ALERT_CONFIG_DIR")}'
+            with safe_path_func(
+                open, alert_config_dir, f"{alert_config_dir}/{self.template}.json", "r"
+            ) as templatefile:
+                template = json.load(templatefile)
+        except Exception as err:
+            raise ConfigurationError(
+                "Template file for alerting payload is either missing or invalid JSON: {}".format(
+                    str(err)
+                )
+            )
+        payload = self._render_template(template)
+        if receiver_config.get("payload_fields") is not None:
+            payload.update(receiver_config.get("payload_fields"))
+        return json.dumps(payload)
+
+    def _render_template(self, template_dict):
+        if isinstance(template_dict, (dict, list)):
+            for key, value in template_dict.items():
+                if isinstance(value, dict):
+                    self._render_template(value)
+                elif isinstance(value, list):
+                    for entry in value:
+                        self._render_template(entry)
+                else:
+                    template_dict[key] = Template(value).render(
+                        self.context, undefined=StrictUndefined
+                    )
+        return template_dict
+
+    def send_alert(self):
+        try:
+            response = requests.post(
+                self.receiver_url, data=self.payload, headers=self.headers
+            )
+            response.raise_for_status()
+            logging.info("sent alert to %s", self.template)
+        except Exception as err:
+            if self.throw_if_alert_sending_fails:
+                raise AlertSendingError(str(err))
+            logging.error(err)
+        return response
+
+
+def _get_headers(receiver_config):
```

Comment:
We earlier had the discussion whether this should live inside the Alert class. [You wrote](https://github.com/sse-secure-systems/connaisseur/pull/70#discussion_r579353353):
> If I put get_headers into the Alert class pylint complains that is can be made a function instead of a method. Now I made it a private function. Is that against best practises?

I'm not sure about the exact error message, but it's probably because a method has a `self` argument, and you were not using it inside the function. For this case, Python has the concept of class methods (depend on the class itself instead of a `self` instance) and static methods (no class- or instance-related argument). To learn more, checkout the `@classmethod` and `@staticmethod` decorators.

Reply:
Cool, thanks! Used it :+1: